DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs

```diff
@@ -105,9 +122,605 @@ internal enum SupportedMarshallingMethods
+                {
+                    methods |= SupportedMarshallingMethods.ManagedToNative;
+                }
+                else if (ManualTypeMarshallingHelper.IsStackallocConstructor(ctor, type, spanOfByte, marshallingVariant)
+                    && (valueProperty is null or { GetMethod: not null }))
+                {
+                    methods |= SupportedMarshallingMethods.ManagedToNativeStackalloc;
+                }
+                else if (ctor.Parameters.Length == 1 && ctor.Parameters[0].Type.SpecialType == SpecialType.System_Int32)
+                {
+                    hasInt32Constructor = true;
+                }
+            }
+
+            // The constructor that takes only the native element size is required for collection marshallers
+            // in the native-to-managed scenario.
+            if ((!isContiguousCollectionMarshaller
+                    || (hasInt32Constructor && ManualTypeMarshallingHelper.HasSetUnmarshalledCollectionLengthMethod(nativeType)))
+                && ManualTypeMarshallingHelper.HasToManagedMethod(nativeType, type)
+                && (valueProperty is null or { SetMethod: not null }))
+            {
+                methods |= SupportedMarshallingMethods.NativeToManaged;
+            }
+
+            if (methods == SupportedMarshallingMethods.None)
+            {
+                diagnostics.ReportConfigurationNotSupported(attrData, "Native Type", nativeType.ToDisplayString());
+                return NoMarshallingInfo.Instance;
+            }
+
+            if (isContiguousCollectionMarshaller)
+            {
+                if (!ManualTypeMarshallingHelper.HasNativeValueStorageProperty(nativeType, spanOfByte))
+                {
+                    diagnostics.ReportConfigurationNotSupported(attrData, "Native Type", nativeType.ToDisplayString());
+                    return NoMarshallingInfo.Instance;
+                }
+
+                if (!ManualTypeMarshallingHelper.TryGetElementTypeFromContiguousCollectionMarshaller(nativeType, out ITypeSymbol elementType))
+                {
+                    diagnostics.ReportConfigurationNotSupported(attrData, "Native Type", nativeType.ToDisplayString());
+                    return NoMarshallingInfo.Instance;
+                }
+
+                return new NativeContiguousCollectionMarshallingInfo(
+                    nativeType,
+                    valueProperty?.Type,
+                    methods,
+                    NativeTypePinnable: ManualTypeMarshallingHelper.FindGetPinnableReference(nativeType) is not null,
+                    UseDefaultMarshalling: !isMarshalUsingAttribute,
+                    parsedCountInfo,
+                    elementType,
+                    GetMarshallingInfo(elementType, useSiteAttributes, maxIndirectionLevelUsed = indirectionLevel + 1, inspectedElements, ref maxIndirectionLevelUsed));
```

Comment:
> `GetMarshallingInfo(elementType, useSiteAttributes, maxIndirectionLevelUsed = indirectionLevel + 1, inspectedElements, ref maxIndirectionLevelUsed)`

Not a fan of this at all. I am sure the evaluation order in C# is defined but it is weird to see how this works coming from C/C++. Let's avoid these sort of inline assignment and uses.